library/pkcs7.c

```diff
@@ -0,0 +1,525 @@
+    signers_set->issuer_raw.p = *p;
+
+    ret = mbedtls_asn1_get_tag( p, end_set, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO + ret );
+
+    ret  = mbedtls_x509_get_name( p, *p + len, &signers_set->issuer );
+    if( ret != 0 )
+        return( ret );
+
+    signers_set->issuer_raw.len =  *p - signers_set->issuer_raw.p;
+
+    ret = mbedtls_x509_get_serial( p, end_set, &signers_set->serial );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_digest_algorithm( p, end_set,
+            &signers_set->alg_identifier );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_digest_algorithm( p, end_set,
+            &signers_set->sig_alg_identifier );
+    if( ret != 0 )
+        return( ret );
+
+    ret = pkcs7_get_signature( p, end, &signers_set->sig );
+    if( ret != 0 )
+        return( ret );
+
+    signers_set->next = NULL;
+
+    return( 0 );
+}
+
+/**
+ * SignedData ::= SEQUENCE {
+ *      version Version,
+ *      digestAlgorithms DigestAlgorithmIdentifiers,
+ *      contentInfo ContentInfo,
+ *      certificates
+ *              [0] IMPLICIT ExtendedCertificatesAndCertificates
+ *                  OPTIONAL,
+ *      crls
+ *              [0] IMPLICIT CertificateRevocationLists OPTIONAL,
+ *      signerInfos SignerInfos }
+ */
+static int pkcs7_get_signed_data( unsigned char *buf, size_t buflen,
+        mbedtls_pkcs7_signed_data *signed_data )
+{
+    unsigned char *p = buf;
+    unsigned char *end = buf + buflen;
+    size_t len = 0;
+    int ret;
+    mbedtls_md_type_t md_alg;
+
+    ret = mbedtls_asn1_get_tag( &p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_FORMAT + ret );
+
+    /* Get version of signed data */
+    ret = pkcs7_get_version( &p, end, &signed_data->version );
+    if( ret != 0 )
+        return( ret );
+
+    /* If version != 1, return invalid version */
+    if( signed_data->version != MBEDTLS_PKCS7_SUPPORTED_VERSION ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_VERSION );
+    }
+
+    /* Get digest algorithm */
+    ret = pkcs7_get_digest_algorithm_set( &p, end,
+            &signed_data->digest_alg_identifiers );
+    if( ret != 0 ) {
+        return( ret );
+    }
+
+    ret = mbedtls_oid_get_md_alg( &signed_data->digest_alg_identifiers, &md_alg );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_ALG + ret );
+
+    /* Do not expect any content */
+    ret = pkcs7_get_content_info_type( &p, end, &signed_data->content.oid );
+    if( ret != 0 )
+        return( ret );
+
+    if( MBEDTLS_OID_CMP( MBEDTLS_OID_PKCS7_DATA, &signed_data->content.oid ) ) {
+        return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO ) ;
+    }
+
+    p = p + signed_data->content.oid.len;
+
+    /* Look for certificates, there may or may not be any */
+    ret = pkcs7_get_next_content_len( &p, end, &len );
```

Comment:
This is difficult to read. I'd suggest you mimick the way the existing X.509 CRT parsing code handles optional fields: Introduce a helper function which processes the optional element if present, and does nothing if it's not present. This way, the top-level parsing routine is easy to read since it has one function call per field, regardless of whether it's optional or not. See e.g. https://github.com/ARMmbed/mbedtls/blob/development/library/x509_crt.c#L452